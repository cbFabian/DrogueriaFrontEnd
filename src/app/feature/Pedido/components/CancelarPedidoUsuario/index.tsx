import * as PropTypes from 'prop-types';
import * as React from 'react';
import { Button } from 'app/shared/components/Button';
import { PedidoListar } from 'app/feature/Pedido/models/PedidoListar';

interface BtnCancelarPedidoUsuarioProps {
    onCancelar: (pedidoListar: PedidoListar) => any;
    pedidoListar: PedidoListar;
}

export const BtnCancelarPedidoUsuario: React.FC<BtnCancelarPedidoUsuarioProps> = ({
    onCancelar,
    pedidoListar,
}) => {
    const handleCancelar = () => onCancelar(pedidoListar);
    return(
        <Button onClick={handleCancelar}>
            <span role='img' aria-labelledby='cancelar'>
                Cancelar‚ùå
            </span>
        </Button>
    );
};

BtnCancelarPedidoUsuario.propTypes = {
    pedidoListar: PropTypes.shape({
        id: PropTypes.number.isRequired,
        nombreUsuario: PropTypes.string.isRequired,
        nombreProducto: PropTypes.string.isRequired,
        tipoReunion: PropTypes.string.isRequired,
        fechaRealizacion: PropTypes.string.isRequired,
        direccion: PropTypes.string.isRequired,
        valorTotal: PropTypes.number.isRequired,
        horasDeServicio: PropTypes.number.isRequired,
    }).isRequired,
    onCancelar:PropTypes.func.isRequired,
};
