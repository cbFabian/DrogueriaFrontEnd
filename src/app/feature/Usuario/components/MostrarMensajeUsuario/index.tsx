import * as PropTypes from 'prop-types';
import * as React from 'react';
import { H2MensajeExito } from './styles';

interface MostrarMensajeUsuarioProps {
  mensaje: string;
}

export const MostrarMensajeUsuario: React.FC<MostrarMensajeUsuarioProps> = ({
    mensaje,
}) => {
  if (mensaje === '') {
    return null;
  }

  return (
    <div>
        <H2MensajeExito>
            {mensaje}
        </H2MensajeExito>
    </div>
  );
};

MostrarMensajeUsuario.propTypes = {
  mensaje: PropTypes.string.isRequired,
};
