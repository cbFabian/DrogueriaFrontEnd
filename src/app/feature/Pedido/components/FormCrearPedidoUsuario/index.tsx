import * as PropTypes from 'prop-types';
import * as React from 'react';
import * as Yup from 'yup';
import { Button } from 'app/shared/components/Button';
import { FormikHelpers } from 'formik/dist/types';
import { Input } from 'app/shared/components/Input';
import { Pedido } from '../../../Pedido/models/Pedido';
import { SpanError, Select, H2EsFestivo } from './styles';
import { useFormik } from 'formik';
import { Producto } from 'app/feature/Producto/models/Producto';
import { Usuario } from 'app/feature/Usuario/models/Usuario';
import { Reunion } from 'app/feature/Reunion/models/Reunion';
import DatePicker from 'react-datepicker';
import 'react-datepicker/dist/react-datepicker.css';
import { useState, useEffect } from 'react';
import { MostrarMensaje } from '../MostrarMensaje/index';
import { MostrarMensajeError } from '../MostrarMensajeError';
import { constantes } from 'app/shared/utils/constantes';
import { calcularValores } from 'app/shared/utils/calcularValores';

const {
    HORAS24,
    HORA15,
    HORA19,
    MINUTOS60,
    SEGUNDOS60,
    MILISEGUNDOS,
    HORAMINIMA,
    HORAMAXIMA,
} = constantes;
const fechaDeHoy: Date = new Date();
const fechaDiaSiguiente: Date = new Date(fechaDeHoy.getTime() + (HORAS24 * MINUTOS60 * SEGUNDOS60 * MILISEGUNDOS));
const fechaYHoraInicial: Date = new Date(fechaDiaSiguiente.setHours(HORA15,0,0));
const fechaYHoraMax: Date = new Date(fechaDiaSiguiente.setHours(HORA19,0,0));

interface FormValues {
    producto: string;
    reunion: string;
    fechaRealizacion: string;
    direccion: string;
    valorTotal: number;
    horasDeServicio: number;
}

interface FormCrearPedidoUsuarioProp {
  onSubmit: (payload: Pedido) => any;
  validarDiaFestivo: (fechaFestivo: Date) => void;
  disabled?: boolean;
  formTitle: string;
  productos: Producto[];
  usuarioPedido: Usuario;
  reuniones: Reunion[];
  mensajePedido: string;
  mensajeExcepcion: string;
  esFestivo: boolean;
  initialValues?: FormValues;
}

const validationSchema = Yup.object().shape<FormValues>({
    producto: Yup.string().required('El Producto es requerido.'),
    reunion: Yup.string().required('La Reunion es requerida.'),
    fechaRealizacion: Yup.string().required('La Fecha  de Realizacion es requerida.'),
    direccion: Yup.string().required('El campo Direccion es requerido.'),
    valorTotal: Yup.number().required('El campo Valor Total es requerido.')
    .positive('El valor del pedido no puede ser 0'),
    horasDeServicio: Yup.number().required('El campo Horas de Servicio es requerido.')
    .positive().integer()
    .min(HORAMINIMA, 'Minimo 4 horas')
    .max(HORAMAXIMA, 'Maximo 8 horas'),
});

export const FormCrearPedidoUsuario: React.FC<FormCrearPedidoUsuarioProp> = ({
    onSubmit,
    validarDiaFestivo,
    disabled,
    formTitle,
    usuarioPedido,
    productos,
    reuniones,
    mensajePedido,
    mensajeExcepcion,
    esFestivo,
    initialValues = {
        producto: JSON.stringify(productos[0]),
        reunion: JSON.stringify(reuniones[0]),
        fechaRealizacion: '',
        direccion: '',
        valorTotal: 0,
        horasDeServicio: 0,
  },
}) => {
  const handleSubmit = (
    values: FormValues,
    { resetForm }: FormikHelpers<FormValues>
) => {
        onSubmit({
            usuario: usuarioPedido,
            producto: JSON.parse(values.producto),
            reunion: JSON.parse(values.reunion),
            fechaRealizacion: fechaInicio.toISOString(),
            direccion: values.direccion,
            valorTotal: calcularValores(values.producto, values.reunion, esFestivo),
            horasDeServicio: values.horasDeServicio,
        });
        resetForm();
    };
    const formik = useFormik({
        initialValues,
        validationSchema,
        onSubmit:handleSubmit,
    });
    let valor = calcularValores(formik.values.producto, formik.values.reunion, esFestivo);
    formik.values.valorTotal = valor;
    const [fechaInicio, setfechaInicio] = useState(new Date(fechaDiaSiguiente.setHours(HORA15,0,0)));
    formik.values.fechaRealizacion = !fechaInicio
        ? ''
        : fechaInicio.toString();
    useEffect(() => {
        validarDiaFestivo(fechaInicio);
      },[validarDiaFestivo, fechaInicio]);
    return(
        <form onSubmit= {formik.handleSubmit}>
            <h2>{formTitle}</h2>
            <MostrarMensaje 
                mensaje={mensajePedido}
            />
            <MostrarMensajeError
                mensajeError={mensajeExcepcion} 
            />
            {esFestivo && 
                <H2EsFestivo>
                    El día seleccionado es festivo, se cobra el doble
                </H2EsFestivo>}
            <label>
                Productos Disponibles:{' '}
            </label>
            <Select
                value={formik.values.producto}
                name="producto"
                onChange={formik.handleChange}>
                {productos.map((producto) => (
                    <option 
                        key={producto.nombre} 
                        value={JSON.stringify(producto)}
                        >
                            {producto.nombre}
                    </option>
                ))}
            </Select>
            {formik.touched.producto && formik.errors.producto && (
                <SpanError>{formik.errors.producto}</SpanError>
            )}              
            <label>
                Tipos de Reuniones:{' '}
            </label>
            <Select
                 value={formik.values.reunion}
                name="reunion"  
                onChange={formik.handleChange}>
                {reuniones.map((reunion) => (
                    <option 
                        key={reunion.tipo} 
                        value={JSON.stringify(reunion)}
                        >
                            {reunion.tipo}
                    </option>
                ))}
            </Select>
            {formik.touched.reunion && formik.errors.reunion && (
                <SpanError>{formik.errors.reunion}</SpanError>
            )}
            <label>
                Fecha y hora de inicio del evento:
            </label>
            <DatePicker
                name="fechaRealizacion" 
                selected={fechaInicio}
                onChange={(date) => {!date ? new Date(): setfechaInicio(date);}}
                showTimeSelect
                minTime={fechaYHoraInicial}
                maxTime={fechaYHoraMax}
                minDate={fechaYHoraInicial}
                dateFormat="MMMM d, yyyy h:mm aa"
            />
            {formik.touched.fechaRealizacion && formik.errors.fechaRealizacion && (
                <SpanError>{formik.errors.fechaRealizacion}</SpanError>
            )}                
            <Input 
                type="hidden"
                name="fechaRealizacionAux"
                value={formik.values.fechaRealizacion}
                onChange={formik.handleChange}
            />
            <Input
                disabled={disabled}
                name="direccion"
                placeholder="Ingrese la Dirección del evento. (Solo en Medellín)"
                value={formik.values.direccion}
                onChange={formik.handleChange} 
            />
            {formik.touched.direccion && formik.errors.direccion && (
                <SpanError>{formik.errors.direccion}</SpanError>
            )}
            <Input
                type="number"
                disabled={disabled}
                name="horasDeServicio"
                placeholder="Ingrese el numero Horas del servicio, (minimo 4 horas)"
                value={formik.values.horasDeServicio}
                onChange={formik.handleChange} 
            />
            {formik.touched.horasDeServicio && formik.errors.horasDeServicio && (
                <SpanError>{formik.errors.horasDeServicio}</SpanError>
            )}
            <label 
                htmlFor="valorTotalPedido">
                    Valor total de su pedido:{formik.values.valorTotal}
            </label>
            {formik.touched.valorTotal && formik.errors.valorTotal && (
                <SpanError>{formik.errors.valorTotal}</SpanError>
            )}            
            <Button type="submit">Hacer Pedido</Button>
        </form>
    );
};

FormCrearPedidoUsuario.propTypes = {
    onSubmit: PropTypes.func.isRequired,
    validarDiaFestivo: PropTypes.func.isRequired,
    formTitle: PropTypes.string.isRequired,
    disabled: PropTypes.bool,
    initialValues: PropTypes.shape({
        producto: PropTypes.string.isRequired,
        reunion: PropTypes.string.isRequired,
        fechaRealizacion: PropTypes.string.isRequired,
        direccion: PropTypes.string.isRequired,
        valorTotal: PropTypes.number.isRequired,
        horasDeServicio: PropTypes.number.isRequired,
    }),
};
