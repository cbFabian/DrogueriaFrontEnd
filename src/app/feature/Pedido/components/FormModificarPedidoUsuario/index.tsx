import * as PropTypes from 'prop-types';
import * as React from 'react';
import * as Yup from 'yup';
import { Button } from 'app/shared/components/Button';
import { FormikHelpers } from 'formik/dist/types';
import { Input } from 'app/shared/components/Input';
import { Pedido } from '../../../Pedido/models/Pedido';
import { PedidoListar } from '../../../Pedido/models/PedidoListar';
import { SpanError, Select } from './styles';
import { useFormik } from 'formik';
import { Producto } from 'app/feature/Producto/models/Producto';
import { Usuario } from 'app/feature/Usuario/models/Usuario';
import { Reunion } from 'app/feature/Reunion/models/Reunion';
import DatePicker from 'react-datepicker';
import 'react-datepicker/dist/react-datepicker.css';
import { useState } from 'react';
import { MostrarMensaje } from '../MostrarMensaje';
import { constantes } from 'app/shared/utils/constantes';
import { calcularValores } from 'app/shared/utils/calcularValores';
import { MostrarMensajeError } from '../MostrarMensajeError';
import { Link } from 'react-router-dom';

const {
    HORAS24,
    HORA15,
    HORA19,
    MINUTOS60,
    SEGUNDOS60,
    MILISEGUNDOS,
    HORAMINIMA,
    HORAMAXIMA,
} = constantes;
const fechaDeHoy: Date = new Date();
const fechaDiaSiguiente: Date = new Date(fechaDeHoy.getTime() + (HORAS24 * MINUTOS60 * SEGUNDOS60 * MILISEGUNDOS));
const fechaYHoraInicialModif: Date = new Date(fechaDiaSiguiente.setHours(HORA15,0,0));
const fechaYHoraMaxModif: Date = new Date(fechaDiaSiguiente.setHours(HORA19,0,0));

interface FormValues {
    producto: string;
    reunion: string;
    fechaRealizacion: string;
    direccion: string;
    valorTotal: number;
    horasDeServicio: number;
}

interface FormModificarPedidoUsuarioProp {
  onSubmit: (pedidoListar: PedidoListar, payload: Pedido) => any;
  disabled?: boolean;
  formTitle: string;
  productos: Producto[];
  usuarios: Usuario[];
  reuniones: Reunion[];
  pedidoListar: PedidoListar;
  mensajeModificar: string;
  mensajeExcepcion: string;
  esFestivo: boolean;
  initialValues?: FormValues;
}

const validationSchema = Yup.object().shape<FormValues>({
    producto: Yup.string().required('El Producto es requerido.'),
    reunion: Yup.string().required('La Reunion es requerida.'),
    fechaRealizacion: Yup.string().required('La Fecha  de Realizacion es requerida.'),
    direccion: Yup.string().required('El campo Direccion es requerido.'),
    valorTotal: Yup.number().required('El campo Valor Total es requerido.')
    .positive('El valor del pedido no puede ser 0'),
    horasDeServicio: Yup.number().required('El campo Horas de Servicio es requerido.')
    .positive().integer()
    .min(HORAMINIMA, 'Minimo 4 horas')
    .max(HORAMAXIMA, 'Maximo 8 horas'),
});

export const FormModificarPedidoUsuario: React.FC<FormModificarPedidoUsuarioProp> = ({
    onSubmit,
    disabled,
    formTitle,
    usuarios,
    productos,
    reuniones,
    pedidoListar,
    mensajeModificar,
    mensajeExcepcion,
    esFestivo,
    initialValues = {
        producto: '',
        reunion: '',
        fechaRealizacion: '',
        direccion: '',
        valorTotal: 0,
        horasDeServicio: 0,
  },
}) => {
  const handleSubmit = (
    values: FormValues,
    { resetForm }: FormikHelpers<FormValues>
) => {
        onSubmit(pedidoListar, {
            usuario: usuarios[0],
            producto: JSON.parse(values.producto),
            reunion: JSON.parse(values.reunion),
            fechaRealizacion: new Date(fechaInicio).toISOString(),
            direccion: values.direccion,
            valorTotal: calcularValores(values.producto, values.reunion, esFestivo),
            horasDeServicio: values.horasDeServicio,
        });
        resetForm();
    };
    const formik = useFormik({
        initialValues,
        validationSchema,
        onSubmit:handleSubmit,
    });
    let valor = calcularValores(formik.values.producto, formik.values.reunion, esFestivo);
    formik.values.valorTotal = valor;
    const [fechaInicio, setfechaInicio] = useState(new Date(fechaDiaSiguiente.setHours(HORA15,0,0)));
    const fechaRealizacion = (fecha: Date) => {
         formik.values.fechaRealizacion = !fecha
        ? new Date(fechaDiaSiguiente).toString()
        : fecha.toString();
        return formik.values.fechaRealizacion;
    };
    return(
        <form onSubmit= {formik.handleSubmit}>
            <h2>{formTitle}</h2>
            <MostrarMensaje 
                mensaje={mensajeModificar}
            />
            <MostrarMensajeError
                mensajeError={mensajeExcepcion} 
            />
            <label>
            Modifique el Producto{' '}
            </label>
            <Select
                name="producto"
                onChange={formik.handleChange}>
                {productos.map((producto) => (
                    <option 
                        key={producto.nombre} 
                        value={JSON.stringify(producto)}
                        >
                            {producto.nombre}
                    </option>
                ))}
            </Select>
            {formik.touched.producto && formik.errors.producto && (
                <SpanError>{formik.errors.producto}</SpanError>
            )}
            <label>
                Cambie el Tipo de Reunión:{' '}
            </label>
            <Select 
                name="reunion"  
                onChange={formik.handleChange}>
                {reuniones.map((reunion) => (
                    <option 
                        key={reunion.tipo} 
                        value={JSON.stringify(reunion)}
                        >
                            {reunion.tipo}
                    </option>
                ))}
            </Select>
            {formik.touched.reunion && formik.errors.reunion && (
                <SpanError>{formik.errors.reunion}</SpanError>
            )}
            <label>
                Modifique la fecha y hora del evento:
            </label>
            <DatePicker
                name="fechaRealizacion" 
                selected={fechaInicio}
                onChange={(date) => {!date ? new Date(): setfechaInicio(date);}}
                showTimeSelect
                minTime={fechaYHoraInicialModif}
                maxTime={fechaYHoraMaxModif}
                minDate={fechaYHoraInicialModif}
                dateFormat="MMMM d, yyyy h:mm aa"
            />
            {formik.touched.fechaRealizacion && formik.errors.fechaRealizacion && (
                <SpanError>{formik.errors.fechaRealizacion}</SpanError>
            )}
            <Input 
                type="hidden"
                name="fechaRealizacionAux"
                value={fechaRealizacion(fechaInicio)}
                onChange={formik.handleChange}
            />
            <Input
                disabled={disabled}
                name="direccion"
                placeholder="Ingrese la Dirección del evento. (Solo en Medellín)"
                value={formik.values.direccion}
                onChange={formik.handleChange} 
            />
            {formik.touched.direccion && formik.errors.direccion && (
                <SpanError>{formik.errors.direccion}</SpanError>
            )}
            <Input
                type="number"
                disabled={disabled}
                name="horasDeServicio"
                placeholder="Ingrese el numero Horas del servicio, (minimo 4 horas)"
                value={formik.values.horasDeServicio}
                onChange={formik.handleChange} 
            />
            {formik.touched.horasDeServicio && formik.errors.horasDeServicio && (
                <SpanError>{formik.errors.horasDeServicio}</SpanError>
            )}
            <label 
                htmlFor="valorTotalPedido">
                    Valor total de su pedido a modificar:{formik.values.valorTotal}
            </label>
            {formik.touched.valorTotal && formik.errors.valorTotal && (
                <SpanError>{formik.errors.valorTotal}</SpanError>
            )}
            <Button type="submit">Modificar Pedido</Button>
            <Link to='/pedidos' replace={true}>
                <Button>
                    Regresar
                </Button>
            </Link>
        </form>
    );
};

FormModificarPedidoUsuario.propTypes = {
    onSubmit: PropTypes.func.isRequired,
    formTitle: PropTypes.string.isRequired,
    disabled: PropTypes.bool,
    initialValues: PropTypes.shape({
        producto: PropTypes.string.isRequired,
        reunion: PropTypes.string.isRequired,
        fechaRealizacion: PropTypes.string.isRequired,
        direccion: PropTypes.string.isRequired,
        valorTotal: PropTypes.number.isRequired,
        horasDeServicio: PropTypes.number.isRequired,
    }),
};
