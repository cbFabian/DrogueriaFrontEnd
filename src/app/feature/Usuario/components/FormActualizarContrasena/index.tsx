import * as PropTypes from 'prop-types';
import * as React from 'react';
import * as Yup from 'yup';
import { Button } from 'app/shared/components/Button';
import { FormikHelpers } from 'formik/dist/types';
import { Input } from 'app/shared/components/Input';
import { CambioClaveUsuario } from '../../models/CambioClaveUsuario';
import { H2Error, SpanError } from './styles';
import { useFormik } from 'formik';
import { Usuario } from '../../models/Usuario';
import { Link } from 'react-router-dom';

const MINIMOCARACTERES = 4;
interface FormValues {
    claveActual: string;
    claveNueva: string;
    confirmarclaveNueva: string;
}

interface FormActualizarContrasenaProp {
    onSubmit: (payload: CambioClaveUsuario) => any;
    borrarMensajeError: () => void;
    disabled?: boolean;
    formTitle: string;
    usuario: Usuario;
    mensajeError: string;
    initialValues?: FormValues;
}

const validationSchema = Yup.object().shape<FormValues>({
    claveActual: Yup.string().required('El campo clave actual es requerido.'),
    claveNueva: Yup.string().required('El campo de nueva clave es requerido.')
    .min(MINIMOCARACTERES, 'Numero minimo de caracteres es 4'),
    confirmarclaveNueva: Yup.string().oneOf([Yup.ref('claveNueva'), undefined], 'Error, La nueva clave no coincide')
    .required('El campo de confirmar nueva clave es requerido.')
    .min(MINIMOCARACTERES, 'Numero minimo de caracteres es 4'),
});

export const FormActualizarContrasena: React.FC<FormActualizarContrasenaProp> = ({
    onSubmit,
    borrarMensajeError,
    disabled,
    formTitle,
    usuario,
    mensajeError,
    initialValues = {
        claveActual: '',
        claveNueva: '',
        confirmarclaveNueva: '',
    },
}) => {
    const handleSubmit = (
        values: FormValues,
        { resetForm }: FormikHelpers<FormValues>
    ) => {
        onSubmit({
            nombre: usuario.nombre,
            claveActual: values.claveActual,
            claveNueva: values.claveNueva,
        });
        resetForm();
    };
    const formik = useFormik({
        initialValues,
        validationSchema,
        onSubmit:handleSubmit,
    });
    if (mensajeError.length > 0) {
        return (
            <div>
                <H2Error>
                    {mensajeError}
                </H2Error>
                <Link to='/ajustes' replace={true}>
                    <Button onClick={borrarMensajeError}>
                        <span role='img' aria-labelledby='regresar'>
                            ⬅️Regresar⬅️
                        </span>
                    </Button>
                </Link>
            </div>
        );
    }
    return(
        <form onSubmit= {formik.handleSubmit}>
            <h2>{formTitle}</h2>
            <Input
                disabled={disabled}
                name="claveActual"
                placeholder="Clave Actual"
                value={formik.values.claveActual}
                onChange={formik.handleChange} 
            />
            {formik.touched.claveActual && formik.errors.claveActual && (
                <SpanError>{formik.errors.claveActual}</SpanError>
            )}
            <Input
                disabled={disabled}
                name="claveNueva"
                placeholder="Ingrese la nueva Clave"
                value={formik.values.claveNueva}
                onChange={formik.handleChange} 
            />
            {formik.touched.claveNueva && formik.errors.claveNueva && (
                <SpanError>{formik.errors.claveNueva}</SpanError>
            )}
            <Input
                disabled={disabled}
                name="confirmarclaveNueva"
                placeholder="Confirme la Nueva Clave"
                value={formik.values.confirmarclaveNueva}
                onChange={formik.handleChange} 
            />
            {formik.touched.confirmarclaveNueva && formik.errors.confirmarclaveNueva && (
                <SpanError>{formik.errors.confirmarclaveNueva}</SpanError>
            )}
            <Button type="submit">Actualizar</Button>
        </form>
    );
};

FormActualizarContrasena.propTypes = {
    onSubmit: PropTypes.func.isRequired,
    formTitle: PropTypes.string.isRequired,
    disabled: PropTypes.bool,
    mensajeError: PropTypes.string.isRequired,
    usuario: PropTypes.shape({
        nombre: PropTypes.string.isRequired,
        clave: PropTypes.string.isRequired,
    }).isRequired,
    initialValues: PropTypes.shape({
        claveActual: PropTypes.string.isRequired,
        claveNueva: PropTypes.string.isRequired,
        confirmarclaveNueva: PropTypes.string.isRequired,
    }),
};
