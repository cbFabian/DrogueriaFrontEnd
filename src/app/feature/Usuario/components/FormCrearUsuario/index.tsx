import * as PropTypes from 'prop-types';
import * as React from 'react';
import * as Yup from 'yup';
import { Button } from 'app/shared/components/Button';
import { FormikHelpers } from 'formik/dist/types';
import { Input } from 'app/shared/components/Input';
import { Usuario } from '../../models/Usuario';
import { SpanError } from './styles';
import { useFormik } from 'formik';
import { MostrarMensajeUsuario } from '../MostrarMensajeUsuario';

const MINIMOCARACTERES = 4;
interface FormValues {
    nombre: string;
    clave: string;
    confirmarClave: string;
}

interface FormCrearUsuarioProp {
    onSubmit: (payload: Usuario) => any;
    irInicioSesion: () => void;
    disabled?: boolean;
    formTitle: string;
    mensajeErrorCreacion: string;
    initialValues?: FormValues;
}

const validationSchema = Yup.object().shape<FormValues>({
    nombre: Yup.string().required('El campo nombre es requerido.'),
    clave: Yup.string().required('El campo clave es requerido.')
    .min(MINIMOCARACTERES, 'Numero minimo de caracteres es 4'),
    confirmarClave: Yup.string().oneOf([Yup.ref('clave'), undefined], 'Error, La clave no coincide')
    .required('El campo de confirmar clave es requerido.')
    .min(MINIMOCARACTERES, 'Numero minimo de caracteres es 4'),
});

export const FormCrearUsuario: React.FC<FormCrearUsuarioProp> = ({
    onSubmit,
    irInicioSesion,
    disabled,
    formTitle,
    mensajeErrorCreacion,
    initialValues = {
        nombre: '',
        clave: '',
        confirmarClave: '',
    },
}) => {
    const handleSubmit = (
        values: FormValues,
        { resetForm }: FormikHelpers<FormValues>
    ) => {
        onSubmit({
            nombre: values.nombre,
            clave: values.clave,
        });
        resetForm();
    };
    const formik = useFormik({
        initialValues,
        validationSchema,
        onSubmit:handleSubmit,
    });
    return(
        <form onSubmit= {formik.handleSubmit}>
            <h2>{formTitle}</h2>
            <MostrarMensajeUsuario 
                mensaje={mensajeErrorCreacion}
            />
            <Input
                disabled={disabled}
                name="nombre"
                placeholder="Nombre"
                value={formik.values.nombre}
                onChange={formik.handleChange} 
            />
            {formik.touched.nombre && formik.errors.nombre && (
                <SpanError>{formik.errors.nombre}</SpanError>
            )}
            <Input
                disabled={disabled}
                name="clave"
                placeholder="Clave"
                value={formik.values.clave}
                onChange={formik.handleChange} 
            />
            {formik.touched.clave && formik.errors.clave && (
                <SpanError>{formik.errors.clave}</SpanError>
            )}
            <Input
                disabled={disabled}
                name="confirmarClave"
                placeholder="Confirme la Clave"
                value={formik.values.confirmarClave}
                onChange={formik.handleChange} 
            />
            {formik.touched.confirmarClave && formik.errors.confirmarClave && (
                <SpanError>{formik.errors.confirmarClave}</SpanError>
            )}
            <Button type="submit">Registrar</Button>
        </form>
    );
};

FormCrearUsuario.propTypes = {
    onSubmit: PropTypes.func.isRequired,
    formTitle: PropTypes.string.isRequired,
    disabled: PropTypes.bool,
    mensajeErrorCreacion: PropTypes.string.isRequired,
    initialValues: PropTypes.shape({
        nombre: PropTypes.string.isRequired,
        clave: PropTypes.string.isRequired,
        confirmarClave: PropTypes.string.isRequired,
    }),
};
