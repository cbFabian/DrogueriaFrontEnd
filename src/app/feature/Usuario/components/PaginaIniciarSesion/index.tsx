import * as PropTypes from 'prop-types';
import * as React from 'react';
import * as Yup from 'yup';
import { Button } from 'app/shared/components/Button';
import { Input } from 'app/shared/components/Input';
import { SpanError } from './styles';
import { FormikHelpers, useFormik } from 'formik';
import { Usuario } from '../../models/Usuario';
import { irAgregarUsuario } from 'app/core/redux/acciones/usuario/UsuarioAcciones';
import { useEffect } from 'react';
import { MostrarMensajeUsuario } from '../MostrarMensajeUsuario';

interface FormValues {
    nombre: string;
    clave: string;
}

interface PaginaIniciarSesionProp {
    onSubmit: (payload: Usuario) => any;
    disabled?: boolean;
    paginaTitle: string;
    mensajeErrorSesion: string;
    initialValues?: FormValues;
}

const validationSchema = Yup.object().shape<FormValues>({
    nombre: Yup.string().required('El campo nombre es requerido.'),
    clave: Yup.string().required('El campo clave es requerido.'),
});

export const PaginaIniciarSesion: React.FC<PaginaIniciarSesionProp> = ({
    onSubmit,
    disabled,
    paginaTitle,
    mensajeErrorSesion,
    initialValues = {
        nombre: '',
        clave: '',
    },
}) => {
    const handleSubmit = (
        values: FormValues,
        { resetForm }: FormikHelpers<FormValues>
    ) => {
        onSubmit({
            nombre: values.nombre,
            clave: values.clave,
        });
        resetForm();
    };
    const formik = useFormik({
        initialValues,
        validationSchema,
        onSubmit:handleSubmit,
    });
    useEffect(() => {
        irAgregarUsuario();
      }, []);
    return(
        <form onSubmit= {formik.handleSubmit}>
            <h2>{paginaTitle}</h2>
            <MostrarMensajeUsuario 
                mensaje={mensajeErrorSesion}
            />
            <Input
                disabled={disabled}
                name="nombre"
                placeholder="Nombre"
                value={formik.values.nombre}
                onChange={formik.handleChange} 
            />
            {formik.touched.nombre && formik.errors.nombre && (
                <SpanError>{formik.errors.nombre}</SpanError>
            )}
            <Input
                type="password"
                disabled={disabled}
                name="clave"
                placeholder="Clave"
                value={formik.values.clave}
                onChange={formik.handleChange} 
            />
            {formik.touched.clave && formik.errors.clave && (
                <SpanError>{formik.errors.clave}</SpanError>
            )}
            <Button type="submit">Siguiente</Button>
        </form>
    );
};

PaginaIniciarSesion.propTypes = {
    onSubmit: PropTypes.func.isRequired,
    paginaTitle: PropTypes.string.isRequired,
    disabled: PropTypes.bool,
    initialValues: PropTypes.shape({
        nombre: PropTypes.string.isRequired,
        clave: PropTypes.string.isRequired,
    }),
};
