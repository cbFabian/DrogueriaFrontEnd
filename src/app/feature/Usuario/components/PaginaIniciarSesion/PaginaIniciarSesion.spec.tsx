import * as React from 'react';
import { RenderResult, fireEvent, render, wait } from '@testing-library/react';
import { SinonStub, stub } from 'sinon';
import { PaginaIniciarSesion } from 'app/feature/Usuario/components/PaginaIniciarSesion';
import { setTextEvent } from 'app/shared/utils/test';

describe('FormCrearUsuario test', () => {
  let componentWrapper: RenderResult;
  let componentProps: React.ComponentProps<typeof PaginaIniciarSesion> & {
    onSubmit: SinonStub;
  };

  beforeEach(() => {
    componentProps = {
      paginaTitle: 'Form test',
      onSubmit: stub(),
      mensajeErrorSesion: '',
    };
    componentWrapper = render(<PaginaIniciarSesion {...componentProps} />);
  });

  it('should match snapshot', () => {
    expect(componentWrapper.container).toMatchSnapshot();
  });

  it('should fail on submit all fields missing', async () => {
    const elem = componentWrapper.container;
    const submitButton = elem.querySelector('button[type="submit"]');
    const DOS = 2;

    await wait(() => {
      submitButton && fireEvent.click(submitButton);
    });
    const spans = elem.querySelectorAll('span');
    expect(spans.length).toBe(DOS);
    expect(spans[0].textContent).toBe('El campo nombre es requerido.');
    expect(spans[1].textContent).toBe('El campo clave es requerido.');
  });

  it('should fail on submit one fields missing', async () => {
    const elem = componentWrapper.container;
    const nombre = elem.querySelector('input[name="nombre"]');
    const submitButton = elem.querySelector('button[type="submit"]');

    await wait(() => {
      nombre && fireEvent.change(nombre, setTextEvent('nombre', 'Lorem'));
    });

    await wait(() => {
      submitButton && fireEvent.click(submitButton);
    });
    const spans = elem.querySelectorAll('span');
    expect(spans.length).toBe(1);
    expect(spans[0].textContent).toBe('El campo clave es requerido.');
  });

  it('should submit', async () => {
    const elem = componentWrapper.container;

    const nombre = elem.querySelector('input[name="nombre"]');
    const clave = elem.querySelector('input[name="clave"]');
    const submitButton = elem.querySelector('button[type="submit"]');

    await wait(() => {
      nombre && fireEvent.change(nombre, setTextEvent('nombre', 'Lorem'));
    });
    await wait(() => {
      clave && fireEvent.change(clave, setTextEvent('clave', '1234'));
    });

    await wait(() => {
      submitButton && fireEvent.click(submitButton);
    });

    const formSubmitted = componentProps.onSubmit.firstCall.args[0];

    expect(formSubmitted.nombre).toBe('Lorem');
    expect(formSubmitted.clave).toBe('1234');
  });
});
