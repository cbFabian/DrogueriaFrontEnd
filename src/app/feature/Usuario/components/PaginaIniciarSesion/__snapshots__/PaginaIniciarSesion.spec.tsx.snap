// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormCrearUsuario test should match snapshot 1`] = `
<div>
  <form>
    <h2>
      Form test
    </h2>
    <input
      class="sc-AxirZ eAukyj"
      name="nombre"
      placeholder="Nombre"
      value=""
    />
    <input
      class="sc-AxirZ eAukyj"
      name="clave"
      placeholder="Clave"
      type="password"
      value=""
    />
    <button
      class="sc-AxjAm kgxTKy"
      type="submit"
    >
      Siguiente
    </button>
  </form>
</div>
`;
